### ISSUE TITLE
Issue with Observable.bufferSliding

### ISSUE THREAD

**Issue by oleg-py** [monix/monix#275]
Using `monix-reactive` 2.0.3, 2.1.0 and 2.1.1 / Scala 2.11.8 / JDK 1.8.0.92 / Windows 10.

I wanted to take last N elements that were signalled by an `Observable` and thought `bufferSliding` is what I needed.

```scala
// In REPL
import scala.concurrent.duration._
import monix.reactive.Observable
import monix.execution.Scheduler.Implicits.global
Observable.interval(1.seconds).bufferSliding(5, 1).foreach(println)
```
I get the following output (reproducible with any duration):
```
List(0, 1, 2, 3, 4)
List(1, 2, 3, 4, 5)
List(5, 6, 7, 8, 9)
List(6, 7, 8, 9, 10)
List(10, 11, 12, 13, 14)
List(11, 12, 13, 14, 15)
```

I expected `bufferSliding` to emit every `count` items (which is what doc says, and it's 1 in that case), but it seems to be emitting once after `count` items, then after `size - count`, then again after `count` and so on.

**Comment by alexandru** [comment]
Hi @oleg-py, the current behavior is consistent with other implementations, like that in Scala's standard collections. For example:

```scala
scala> (0 to 15).sliding(5,1).toList.foreach(println)
Vector(0, 1, 2, 3, 4)
Vector(1, 2, 3, 4, 5)
Vector(2, 3, 4, 5, 6)
Vector(3, 4, 5, 6, 7)
Vector(4, 5, 6, 7, 8)
Vector(5, 6, 7, 8, 9)
Vector(6, 7, 8, 9, 10)
Vector(7, 8, 9, 10, 11)
Vector(8, 9, 10, 11, 12)
Vector(9, 10, 11, 12, 13)
Vector(10, 11, 12, 13, 14)
Vector(11, 12, 13, 14, 15)
```

The documentation says something like:
> *The resulting observable emits buffers every skip items, each containing count items.*

Note that your `count` is 5 and not 1, as `count` is the first parameter, not the second, see: https://goo.gl/97QS8e

Granted, this is a little confusing. But what it says is that for `count=5` and `skip=1` we'll end up emitting a list of 5 items for each element emitted by the source.  Also this in case we have `count=5` and `skip=2` we'll get a list of 5 items for each 2 elements emitted by the source:

```scala
scala> (0 to 15).sliding(5,2).toList.foreach(println)
Vector(0, 1, 2, 3, 4)
Vector(2, 3, 4, 5, 6)
Vector(4, 5, 6, 7, 8)
Vector(6, 7, 8, 9, 10)
Vector(8, 9, 10, 11, 12)
Vector(10, 11, 12, 13, 14)
Vector(12, 13, 14, 15)
```

What this means, if you'll notice, is that the source emits `0, 1` and the sliding observable has to emit a list of 5 for it. It can't do that yet, but internally we are slowly building a sliding buffer keeping `count` items in it. And then the source emits `2,3` and the sliding observable has to emit a list of 5 items for it. Of course, we don't have these items yet, but that's OK, we'll wait. Etc..

I'm not sure if this explanation helps, but think about it like this: internally we keep a sliding buffer of the last `count` elements and we emit that buffer downstream every `skip` events.

**Comment by alexandru** [comment]
@oleg-py sorry, I've first replied without thinking. There's indeed something wrong about it. Let me check and get back to you.

**Comment by oleg-py** [comment]
@alexandru no problem, and thanks for the swift response. Let me know if you need any additional information.

**Comment by pjurczenko** [comment]
The problem is with this line [this](https://github.com/monix/monix/blob/master/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/BufferSlidingOperator.scala#L53) of `BufferSlidingOperator`. Instead of `nextBuffer = ListBuffer.empty[A]` there should be `nextBuffer = nextBuffer.drop(skip)`

**Comment by alexandru** [comment]
Sorry folks for taking so long, problem is now fixed and on master, will deploy a new version soon.

**Comment by alexandru** [comment]
The fix is in the latest deployed version: 2.1.2
